mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,442 @@
+    by Clifford operations.
+
+    Args:
+        non_clifford_ops: A sequence of non-Clifford operations.
+        fraction_non_clifford: The (approximate) fraction of non-Clifford
+            operations in the returned list.
+        method_select: The way in which the non-Clifford gates are selected to
+            be replaced by Clifford gates. Options are 'uniform' or 'gaussian'.
+        method_replace: The way in which selected non-Clifford gates are
+            replaced by Clifford gates. Options are 'uniform', 'gaussian' or
+            'closest'.
+        seed: Seed for sampling.
+        kwargs: Additional options for selection / replacement methods.
+            sigma_select (float): Width of the Gaussian distribution used for
+                ``method_select='gaussian'``.
+            sigma_replace (float): Width of the Gaussian distribution used for
+                ``method_replace='gaussian'``.
+    """
+    sigma_select = kwargs.get("sigma_select", 0.5)
+    sigma_replace = kwargs.get("sigma_replace", 0.5)
+
+    state = np.random.RandomState(seed)
+    seed_select, seed_replace = state.randint(1e8, size=2,)
+    random_state_select = np.random.RandomState(seed_select)
+    random_state_replace = np.random.RandomState(seed_replace)
+
+    # Select (indices of) operations to replace.
+    indices_of_selected_ops = _select(
+        non_clifford_ops,
+        fraction_non_clifford,
+        method_select,
+        sigma_select,
+        random_state_select,
+    )
+
+    # Replace selected operations.
+    clifford_ops = _replace(
+        [non_clifford_ops[i] for i in indices_of_selected_ops],
+        method_replace,
+        sigma_replace,
+        random_state_replace,
+    )
+
+    # Return sequence of (near) Clifford operations.
+    return [
+        clifford_ops.pop(0) if i in indices_of_selected_ops else op
+        for (i, op) in enumerate(non_clifford_ops)
+    ]
+
+
+def _select(
+    non_clifford_ops: Sequence[cirq.ops.Operation],
+    fraction_non_clifford: float,
+    method: str = "uniform",
+    sigma: Optional[float] = 1.0,
+    random_state: Optional[np.random.RandomState] = None,
+) -> List[int]:
+    """Returns indices of non-Clifford operations selected (to be replaced)
+    according to some method.
+
+    Args:
+        non_clifford_ops: Sequence of non-Clifford operations.
+        fraction_non_clifford: fraction of non-Clifford gates to change.
+        method: {'uniform', 'gaussian'} method to use to select Clifford gates
+                to replace.
+        sigma: width of probability distribution used in selection
+                      of non-Clifford gates to replace, only has effect if
+                      method_select = 'gaussian'.
+        random_state: Random state for sampling.
+    """
+    if random_state is None:
+        random_state = np.random
+
+    num_non_cliff = len(non_clifford_ops)
+    num_to_replace = int(fraction_non_clifford * num_non_cliff)
```

Comment:
```suggestion
    num_to_replace = int(round(fraction_non_clifford * num_non_cliff))
```
slightly better approximation